File: SharedLogic/BarcodeProductMapper.swift
import Foundation

/// Maps Open Food Facts API JSON responses to product info. Pure Swift — no
/// Apple frameworks.
///
/// The Open Food Facts API returns JSON at:
///   https://world.openfoodfacts.org/api/v2/product/{barcode}.json
///
/// This module parses that JSON into a structured product result that can be
/// used to populate pantry or grocery list items.

struct ProductLookupResult: Codable, Equatable {
    var barcode: String
    var name: String
    var brand: String
    var category: String
    var quantity: String
    var imageURL: String

    init(
        barcode: String = "",
        name: String = "",
        brand: String = "",
        category: String = "",
        quantity: String = "",
        imageURL: String = ""
    ) {
        self.barcode = barcode
        self.name = name
        self.brand = brand
        self.category = category
        self.quantity = quantity
        self.imageURL = imageURL
    }
}

/// Parses an Open Food Facts API v2 JSON response into a `ProductLookupResult`.
/// Returns nil if the product was not found (status != 1).
func parseOpenFoodFactsResponse(_ jsonData: Data) -> ProductLookupResult? {
    guard let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else {
        return nil
    }
    return parseOpenFoodFactsJSON(json)
}

/// Parses the already-deserialized JSON dictionary.
func parseOpenFoodFactsJSON(_ json: [String: Any]) -> ProductLookupResult? {
    // Check status field
    let status = json["status"] as? Int ?? 0
    guard status == 1 else { return nil }

    let product = json["product"] as? [String: Any] ?? [:]
    let barcode = json["code"] as? String ?? ""

    let name = bestProductName(from: product)
    let brand = product["brands"] as? String ?? ""
    let category = mapOFFCategory(product["categories_tags"] as? [String] ?? [])
    let quantity = product["quantity"] as? String ?? ""
    let imageURL = product["image_front_small_url"] as? String ?? ""

    guard !name.isEmpty else { return nil }

    return ProductLookupResult(
        barcode: barcode,
        name: name,
        brand: brand,
        category: category,
        quantity: quantity,
        imageURL: imageURL
    )
}

/// Picks the best available product name from OFF fields.
func bestProductName(from product: [String: Any]) -> String {
    // Prefer product_name_en, then product_name, then generic_name
    let candidates: [String] = [
        product["product_name_en"] as? String ?? "",
        product["product_name"] as? String ?? "",
        product["generic_name"] as? String ?? "",
    ]
    for candidate in candidates {
        let trimmed = candidate.trimmingCharacters(in: .whitespaces)
        if !trimmed.isEmpty { return trimmed }
    }
    return ""
}

/// Category mapping from Open Food Facts tags to our store-aisle categories.
private let offCategoryMapping: [(keywords: [String], category: String)] = [
    (["fruits", "vegetables", "legumes", "salads"], "Produce"),
    (["dairies", "milks", "cheeses", "yogurts", "eggs", "butters", "creams"], "Dairy"),
    (["meats", "poultry", "beef", "pork", "fish", "seafood", "lamb"], "Meat"),
    (["cereals", "pasta", "rice", "canned", "sauces", "oils", "flour", "sugar", "spices"], "Dry & Canned"),
    (["frozen", "ice-cream"], "Frozen"),
    (["breads", "bakery", "pastries"], "Bakery"),
    (["snacks", "chips", "crackers", "cookies", "candy", "chocolate"], "Snacks"),
    (["beverages", "drinks", "juices", "sodas", "waters", "coffee", "tea"], "Beverages"),
    (["condiments", "dressings", "ketchup", "mustard", "mayonnaise"], "Condiments"),
    (["cleaning", "household", "paper"], "Household"),
]

/// Maps Open Food Facts category tags (like "en:dairies", "en:milks") to a
/// store-aisle category string.
func mapOFFCategory(_ tags: [String]) -> String {
    // OFF tags look like "en:dairies", "en:whole-milks" etc.
    let normalized = tags.map { tag -> String in
        let parts = tag.split(separator: ":")
        return (parts.count > 1 ? String(parts[1]) : tag).lowercased()
    }

    for (keywords, category) in offCategoryMapping {
        for tag in normalized {
            for keyword in keywords {
                if tag.contains(keyword) {
                    return category
                }
            }
        }
    }
    return "Other"
}

/// Formats a product name with brand for display: "Brand Name Product" or
/// just "Product" if brand is empty or already in the name.
func formatProductDisplay(name: String, brand: String) -> String {
    let trimmedBrand = brand.trimmingCharacters(in: .whitespaces)
    guard !trimmedBrand.isEmpty else { return name }
    if name.lowercased().contains(trimmedBrand.lowercased()) { return name }
    return "\(trimmedBrand) \(name)"
}

File: recipe-app-rn/src/lib/barcodeProductMapper.ts
/**
 * Maps Open Food Facts API responses to product info. 1:1 port of
 * `SharedLogic/BarcodeProductMapper.swift` (framework-free). The barcode-scan
 * flow (Phase 5) hits `https://world.openfoodfacts.org/api/v2/product/{barcode}.json`
 * and feeds the JSON here to get a structured product for a grocery/pantry item.
 *
 * In Swift the input was `Data`/`[String: Any]`; in TS it's the already-parsed
 * JSON object (from `response.json()`), so the `Data` variant is dropped.
 */

export type ProductLookupResult = {
  barcode: string;
  name: string;
  brand: string;
  category: string;
  quantity: string;
  imageURL: string;
};

type Json = Record<string, unknown>;

function asString(v: unknown): string {
  return typeof v === 'string' ? v : '';
}

/** Picks the best available product name (product_name_en → product_name → generic_name). */
export function bestProductName(product: Json): string {
  const candidates = [
    asString(product.product_name_en),
    asString(product.product_name),
    asString(product.generic_name),
  ];
  for (const candidate of candidates) {
    const trimmed = candidate.trim();
    if (trimmed.length > 0) return trimmed;
  }
  return '';
}

/** Open Food Facts category tags → our store-aisle categories (order = priority). */
const offCategoryMapping: { keywords: string[]; category: string }[] = [
  { keywords: ['fruits', 'vegetables', 'legumes', 'salads'], category: 'Produce' },
  { keywords: ['dairies', 'milks', 'cheeses', 'yogurts', 'eggs', 'butters', 'creams'], category: 'Dairy' },
  { keywords: ['meats', 'poultry', 'beef', 'pork', 'fish', 'seafood', 'lamb'], category: 'Meat' },
  { keywords: ['cereals', 'pasta', 'rice', 'canned', 'sauces', 'oils', 'flour', 'sugar', 'spices'], category: 'Dry & Canned' },
  { keywords: ['frozen', 'ice-cream'], category: 'Frozen' },
  { keywords: ['breads', 'bakery', 'pastries'], category: 'Bakery' },
  { keywords: ['snacks', 'chips', 'crackers', 'cookies', 'candy', 'chocolate'], category: 'Snacks' },
  { keywords: ['beverages', 'drinks', 'juices', 'sodas', 'waters', 'coffee', 'tea'], category: 'Beverages' },
  { keywords: ['condiments', 'dressings', 'ketchup', 'mustard', 'mayonnaise'], category: 'Condiments' },
  { keywords: ['cleaning', 'household', 'paper'], category: 'Household' },
];

/** Maps OFF category tags (like "en:dairies", "en:whole-milks") to a store-aisle category. */
export function mapOFFCategory(tags: string[]): string {
  const normalized = tags.map((tag) => {
    const parts = tag.split(':');
    return (parts.length > 1 ? parts[1] : tag).toLowerCase();
  });
  for (const { keywords, category } of offCategoryMapping) {
    for (const tag of normalized) {
      for (const keyword of keywords) {
        if (tag.includes(keyword)) return category;
      }
    }
  }
  return 'Other';
}

/**
 * Parses an Open Food Facts API v2 JSON object into a {@link ProductLookupResult}.
 * Returns null if the product was not found (status != 1) or has no name.
 */
export function parseOpenFoodFactsJSON(json: unknown): ProductLookupResult | null {
  if (typeof json !== 'object' || json === null) return null;
  const root = json as Json;
  const status = typeof root.status === 'number' ? root.status : 0;
  if (status !== 1) return null;

  const product = (typeof root.product === 'object' && root.product !== null ? root.product : {}) as Json;
  const barcode = asString(root.code);
  const name = bestProductName(product);
  if (name.length === 0) return null;

  const categoriesTags = Array.isArray(product.categories_tags)
    ? (product.categories_tags.filter((t): t is string => typeof t === 'string'))
    : [];

  return {
    barcode,
    name,
    brand: asString(product.brands),
    category: mapOFFCategory(categoriesTags),
    quantity: asString(product.quantity),
    imageURL: asString(product.image_front_small_url),
  };
}

/** "Brand Name Product", or just "Product" if brand is empty / already in the name. */
export function formatProductDisplay(name: string, brand: string): string {
  const trimmedBrand = brand.trim();
  if (trimmedBrand.length === 0) return name;
  if (name.toLowerCase().includes(trimmedBrand.toLowerCase())) return name;
  return `${trimmedBrand} ${name}`;
}
